### ISSUE TITLE
gazelle:default_visibility directive does not work properly

### ISSUE THREAD

**Issue by vadimberezniker** [bazelbuild/bazel-gazelle#1666]
### What version of gazelle are you using?

0.34.0

### What version of rules_go are you using?

0.42.0

### What version of Bazel are you using?

6.4.0

Simple repro here: https://github.com/vadimberezniker/gazelle-visibility-repro

The `enterprise` package has a `default_visibility` directive. The `enterprise/test` directory has a go file that does not have a BUILD file yet. 

First time you run `bazel run //:gazelle` it will generate this BUILD file:

```
load("@io_bazel_rules_go//go:def.bzl", "go_library")
  
  go_library(
      name = "test",
      srcs = ["test.go"],
      importpath = "github.com/vadimberezniker/repro/enterprise/test",
      visibility = ["//visibility:public"],
  )
```

Second time you run `bazel run //:gazelle` it will add the default_visibility:
```
load("@io_bazel_rules_go//go:def.bzl", "go_library")
  
  go_library(
      name = "test",
      srcs = ["test.go"],
      importpath = "github.com/vadimberezniker/repro/enterprise/test",
      visibility = ["//visibility:public"],
  )
  
  package(default_visibility = ["//enterprise:__subpackages__"])
```

Expectation: the test library should have restricted visibility, not public.

**Comment by sluongng** [comment]
cc: @dnathe4th 

The BUILD file must have already existed for `package()` to be generated. If there is no existing BUILD file, only the `go` language extension would pick up `*.go` files and generate targets with `public` visibility.

Then on the 2nd run, the actual `package()` is generated on the newly created BUILD file.

---

I think the fix here is as follows:

1. Add a doc to ensure that language/bazel/visibility should always be run last, and when it runs, it should consider/detect newly generated BUILD files by language extensions that run prior.

2. Apply "corrective" visibility over existing targets (unless `# keep` is used).

WDYT?

**Comment by dnathe4th** [comment]
Yeah I think (1) is the answer. Let me dig in a bit to see why it isn't detecting the newly created BUILD file (since the example repo does have the extension running after the `go` extension. It is unfortunate there isn't a good way to articulate a dependency between extensions.

(2) brings up a good point I went back and forth with in consulting another engineer when I originally authored the extension. There was a reasonable argument to be made that the package level default construct was not needed at all; the directive could just write to all rules' visibility attributes directly. The argument against this approach was that `package(default_visibility)` was a construct that already existed and was not going away, so better to have gazelle lean in on existing bazel rather than layering on an independent path to achieve the same goal. I think the same argument can also be made against overwriting any visibility directly: there's already a bazel pattern of the default and the rule specific attribute co-existing, let's not violate that by making a guess as to whether this extension should rewrite.

**Comment by kfcss** [comment]
`gazelle:default_visibility` still does not work. Is there any plan to actually get it working?
